### ISSUE TITLE
thinking_token_budget rejected on V2 model runner (''not yet supported'') — port budget logits processor for reasoning-parser models

### ISSUE THREAD

**Issue by scottcha** [vllm-project/vllm#50473]
## Summary

`thinking_token_budget` (added in #20859) hard-rejects on the V2 model runner with:

```
thinking_token_budget is not yet supported by the V2 model runner.
Run vLLM with VLLM_USE_V2_MODEL_RUNNER=0 to use thinking_token_budget.
```

Request: port the thinking-budget logits processor (force-close the reasoning phase after N reasoning tokens, via the registered reasoning parser) to the V2 model runner.

## Motivation

We serve Moonshot Kimi-K3 (hybrid KDA/linear-attention arch — `IsHybrid` + `HasInnerState`, separate mamba-cache state) on the official `vllm/vllm-openai:kimi-k3` image, which defaults to V2. The model's default reasoning effort (`thinking_effort=max` when the client omits it) can spend the entire output budget inside the reasoning phase, producing `finish_reason=length` responses with reasoning-only output and zero client-visible content. On our production gateway this terminal class ("reasoning_only empty output") occurs at ~27 per 10K requests on K3 vs ~1.3 per 10K on comparable thinking models.

`thinking_token_budget` is exactly the right governor and works for our non-V2 models (Kimi K2.6, GLM-5.1/5.2) — it force-closes reasoning so content always gets budget. It is unavailable precisely on the one model in our fleet that most needs it.

The V1-runner workaround (`VLLM_USE_V2_MODEL_RUNNER=0`) is likely a non-starter for hybrid/mamba architectures like K3, so a V1 fallback is not a real remediation path here.

## Proposed scope

Support `thinking_token_budget` (chat + completions protocols) under the V2 runner for models with a registered `--reasoning-parser`, with the same semantics as the V1 implementation: inject the reasoning-end token when the budget is reached, preserving the remaining max_tokens budget for content.

## Environment

- Image: `vllm/vllm-openai:kimi-k3` (official K3 release image, vLLM ≥ 0.26 line)
- Model: `moonshotai/Kimi-K3` (mxfp4, TP8, `--reasoning-parser kimi_k3`, `--tool-call-parser kimi_k3`)
- Repro: any `/v1/chat/completions` request with `"thinking_token_budget": 8000` → 400 `BadRequestError` as above

**Comment by chaunceyjiang** [comment]
It doesn’t seem to be supported yet. I’ll continue investigating and work on improving it after this PR https://github.com/vllm-project/vllm/pull/46727 is merged.
